File: services/community-service/src/models/community.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use utoipa::ToSchema;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema, PartialEq)]
pub struct GeoPoint {
    pub lat: f64,
    pub lng: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum CoverageArea {
    Circle {
        center: GeoPoint,
        radius: f64, // in meters
    },
    Polygon {
        coordinates: Vec<GeoPoint>,
    },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum CommunityType {
    Zone,
    Neighborhood,
    Guild,
    StudyGroup,
    /// A container community that groups other communities together
    Group,
}

impl sqlx::Type<sqlx::Postgres> for CommunityType {
    fn type_info() -> sqlx::postgres::PgTypeInfo {
        sqlx::postgres::PgTypeInfo::with_name("varchar")
    }
}

impl<'r> sqlx::Decode<'r, sqlx::Postgres> for CommunityType {
    fn decode(
        value: sqlx::postgres::PgValueRef<'r>,
    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let s: &str = <&str as sqlx::Decode<sqlx::Postgres>>::decode(value)?;
        match s {
            "zone" => Ok(CommunityType::Zone),
            "neighborhood" => Ok(CommunityType::Neighborhood),
            "guild" => Ok(CommunityType::Guild),
            "study_group" => Ok(CommunityType::StudyGroup),
            "group" => Ok(CommunityType::Group),
            _ => Err(format!("Unknown community type: {}", s).into()),
        }
    }
}

impl<'q> sqlx::Encode<'q, sqlx::Postgres> for CommunityType {
    fn encode_by_ref(
        &self,
        buf: &mut sqlx::postgres::PgArgumentBuffer,
    ) -> Result<sqlx::encode::IsNull, Box<dyn std::error::Error + Send + Sync + 'static>> {
        let s = match self {
            CommunityType::Zone => "zone",
            CommunityType::Neighborhood => "neighborhood",
            CommunityType::Guild => "guild",
            CommunityType::StudyGroup => "study_group",
            CommunityType::Group => "group",
        };
        <&str as sqlx::Encode<sqlx::Postgres>>::encode_by_ref(&s, buf)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct Community {
    pub id: Uuid,
    pub slug: String,
    pub name: String,
    pub description: Option<String>,
    #[sqlx(rename = "type")]
    #[serde(rename = "type")]
    pub community_type: CommunityType,
    pub territory_id: Option<String>,
    pub parent_community_id: Option<Uuid>,
    pub avatar_url: Option<String>,
    pub banner_url: Option<String>,
    pub location_lat: Option<f64>,
    pub location_lng: Option<f64>,
    #[schema(value_type = Option<CoverageArea>)]
    pub coverage_area: Option<sqlx::types::Json<serde_json::Value>>,
    pub member_count: i32,
    pub created_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CommunityRole {
    Admin,
    Moderator,
    Member,
}

impl sqlx::Type<sqlx::Postgres> for CommunityRole {
    fn type_info() -> sqlx::postgres::PgTypeInfo {
        <String as sqlx::Type<sqlx::Postgres>>::type_info()
    }
}

impl<'q> sqlx::Encode<'q, sqlx::Postgres> for CommunityRole {
    fn encode_by_ref(
        &self,
        buf: &mut sqlx::postgres::PgArgumentBuffer,
    ) -> Result<sqlx::encode::IsNull, Box<dyn std::error::Error + Send + Sync>> {
        let s = match self {
            CommunityRole::Admin => "admin",
            CommunityRole::Moderator => "moderator",
            CommunityRole::Member => "member",
        };
        <String as sqlx::Encode<sqlx::Postgres>>::encode_by_ref(&s.to_string(), buf)
    }
}

impl<'r> sqlx::Decode<'r, sqlx::Postgres> for CommunityRole {
    fn decode(
        value: sqlx::postgres::PgValueRef<'r>,
    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
        let s: String = <String as sqlx::Decode<sqlx::Postgres>>::decode(value)?;
        match s.as_str() {
            "admin" => Ok(CommunityRole::Admin),
            "moderator" => Ok(CommunityRole::Moderator),
            "member" => Ok(CommunityRole::Member),
            _ => Err(format!("Invalid community role: {}", s).into()),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct CommunityMember {
    pub community_id: Uuid,
    pub user_id: Uuid,
    pub joined_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct CommunitySettings {
    pub community_id: Uuid,
    pub inherit_requirements: bool,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum RequirementContext {
    View,
    Participate,
    Admin,
}

impl sqlx::Type<sqlx::Postgres> for RequirementContext {
    fn type_info() -> sqlx::postgres::PgTypeInfo {
        sqlx::postgres::PgTypeInfo::with_name("varchar")
    }
}

impl<'q> sqlx::Encode<'q, sqlx::Postgres> for RequirementContext {
    fn encode_by_ref(
        &self,
        buf: &mut sqlx::postgres::PgArgumentBuffer,
    ) -> Result<sqlx::encode::IsNull, Box<dyn std::error::Error + Send + Sync>> {
        let s = match self {
            RequirementContext::View => "view",
            RequirementContext::Participate => "participate",
            RequirementContext::Admin => "admin",
        };
        <String as sqlx::Encode<sqlx::Postgres>>::encode_by_ref(&s.to_string(), buf)
    }
}

impl<'r> sqlx::Decode<'r, sqlx::Postgres> for RequirementContext {
    fn decode(
        value: sqlx::postgres::PgValueRef<'r>,
    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
        let s: String = <String as sqlx::Decode<sqlx::Postgres>>::decode(value)?;
        match s.as_str() {
            "view" => Ok(RequirementContext::View),
            "participate" => Ok(RequirementContext::Participate),
            "admin" => Ok(RequirementContext::Admin),
            _ => Err(format!("Invalid requirement context: {}", s).into()),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct CommunityBadgeRequirement {
    pub id: Uuid,
    pub community_id: Uuid,
    pub badge_id: Uuid,
    pub context: RequirementContext,
    pub created_at: DateTime<Utc>,
}

/// A badge requirement with inheritance information
#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct EffectiveBadgeRequirement {
    pub id: Uuid,
    pub community_id: Uuid,
    pub badge_id: Uuid,
    pub context: RequirementContext,
    pub created_at: DateTime<Utc>,
    /// Name of the badge
    pub badge_name: String,
    /// Slug of the badge
    pub badge_slug: String,
    /// Name of the community that defines this requirement
    pub source_community_name: String,
    /// Whether this requirement is inherited from a parent community
    pub is_inherited: bool,
    /// How many levels up this requirement comes from (0 = direct, 1 = parent, etc.)
    pub inheritance_depth: i32,
}

#[derive(Debug, Serialize, Deserialize, utoipa::ToSchema, sqlx::FromRow)]
pub struct EffectiveManager {
    pub user_id: Uuid,
    pub username: String,
    pub avatar_url: Option<String>,
    pub role: String,
    pub source: String,
    pub distance: i32,
}

File: app/src/api/community.ts
import { apiClient } from '@/lib/api-client'

export const CommunityType = {
  Zone: 'zone',
  Neighborhood: 'neighborhood',
  Guild: 'guild',
  StudyGroup: 'study_group',
  /** A container community that groups other communities together */
  Group: 'group',
} as const

export type CommunityType = (typeof CommunityType)[keyof typeof CommunityType]

export interface GeoPoint {
  lat: number
  lng: number
}

export interface CoverageArea {
  type: 'circle' | 'polygon'
  center?: GeoPoint
  radius?: number // in meters
  coordinates?: GeoPoint[]
}

export interface Community {
  id: string
  slug: string
  name: string
  description?: string
  type: CommunityType
  territory_id?: string
  parent_community_id?: string
  avatar_url?: string
  banner_url?: string
  location_lat?: number
  location_lng?: number
  coverage_area?: CoverageArea
  is_public: boolean
  member_count: number
  created_at: string
  updated_at: string
  requirements?: CommunityBadgeRequirement[]
}

export interface CreateCommunityRequest {
  name: string
  description?: string
  community_type: CommunityType
  territory_id?: string
  parent_community_id?: string
  avatar_url?: string
  banner_url?: string
  location_lat?: number
  location_lng?: number
  coverage_area?: CoverageArea
  inherit_requirements?: boolean
  initial_requirements?: AddRequirementRequest[]
}

export interface UpdateCommunityRequest {
  name?: string
  description?: string
  parent_community_id?: string
  avatar_url?: string
  banner_url?: string
  location_lat?: number
  location_lng?: number
  coverage_area?: CoverageArea
}

export interface CommunityFilter {
  community_type?: CommunityType
  parent_id?: string
  territory_id?: string
  search?: string
}

export const RequirementContext = {
  View: 'view',
  Participate: 'participate',
  Admin: 'admin',
} as const

export type RequirementContext = (typeof RequirementContext)[keyof typeof RequirementContext]

export interface CommunityBadgeRequirement {
  id: string
  community_id: string
  badge_id: string
  context: RequirementContext
  created_at: string
}

/** A badge requirement with inheritance information */
export interface EffectiveBadgeRequirement {
  id: string
  communityId: string
  badgeId: string
  context: RequirementContext
  createdAt: string
  /** Name of the badge */
  badgeName: string
  /** Slug of the badge */
  badgeSlug: string
  /** Name of the community that defines this requirement */
  sourceCommunityName: string
  /** Whether this requirement is inherited from a parent community */
  isInherited: boolean
  /** How many levels up this requirement comes from (0 = direct, 1 = parent, etc.) */
  inheritanceDepth: number
}

export interface AddRequirementRequest {
  badge_id: string
  context: RequirementContext
}

export interface EffectiveManager {
  user_id: string
  username: string
  avatar_url?: string
  role: string
  source: string
  distance: number
}

/** Summary of child groups (Guilds and Study Groups) under a community */
export interface GroupSummary {
  communityId: string
  guildCount: number
  studyGroupCount: number
  totalGuildCount: number
  totalStudyGroupCount: number
  totalMembers: number
  hasBadgeRequirement: boolean
  badgeName?: string
  badgeId?: string
  children: GroupChild[]
}

/** A child group community summary */
export interface GroupChild {
  id: string
  name: string
  slug: string
  communityType: CommunityType
  memberCount: number
  hasBadgeRequirement: boolean
  badgeName?: string
}

export const communityService = {
  listCommunities: async (filter: CommunityFilter): Promise<Community[]> => {
    const response = await apiClient.get('/api/v1/communities', { params: filter })
    return response.data
  },

  createCommunity: async (data: CreateCommunityRequest): Promise<Community> => {
    const response = await apiClient.post('/api/v1/communities', data)
    return response.data
  },

  getCommunity: async (id: string): Promise<Community> => {
    const response = await apiClient.get(`/api/v1/communities/${id}`)
    return response.data
  },

  addRequirement: async (communityId: string, data: AddRequirementRequest): Promise<void> => {
    await apiClient.post(`/api/v1/communities/${communityId}/requirements`, data)
  },

  removeRequirement: async (communityId: string, badgeId: string, context: RequirementContext): Promise<void> => {
    await apiClient.delete(`/api/v1/communities/${communityId}/requirements/${badgeId}`, {
      params: { context },
    })
  },

  listRequirements: async (communityId: string): Promise<CommunityBadgeRequirement[]> => {
    const response = await apiClient.get(`/api/v1/communities/${communityId}/requirements`)
    return response.data
  },

  assignManager: async (communityId: string, userId: string): Promise<void> => {
    await apiClient.post(`/api/v1/communities/${communityId}/manager`, { user_id: userId })
  },

  revokeManager: async (communityId: string, userId: string): Promise<void> => {
    await apiClient.delete(`/api/v1/communities/${communityId}/manager/${userId}`)
  },

  updateCommunity: async (id: string, data: UpdateCommunityRequest): Promise<Community> => {
    const response = await apiClient.put(`/api/v1/communities/${id}`, data)
    return response.data
  },

  getEffectiveManagers: async (id: string): Promise<EffectiveManager[]> => {
    const response = await apiClient.get(`/api/v1/communities/${id}/managers`)
    return response.data
  },

  getManagedCommunities: async (): Promise<string[]> => {
    const response = await apiClient.get('/api/v1/communities/managed-by-me')
    return response.data
  },

  joinCommunity: async (id: string): Promise<void> => {
    await apiClient.post(`/api/v1/communities/${id}/join`)
  },

  leaveCommunity: async (id: string): Promise<void> => {
    await apiClient.post(`/api/v1/communities/${id}/leave`)
  },

  getMembership: async (id: string): Promise<{ is_member: boolean }> => {
    const response = await apiClient.get(`/api/v1/communities/${id}/membership`)
    return response.data
  },

  /** Get a summary of child groups (Guilds and Study Groups) under a community */
  getGroupSummary: async (communityId: string): Promise<GroupSummary> => {
    const response = await apiClient.get(`/api/v1/communities/${communityId}/group-summary`)
    return response.data
  },

  /** Get effective badge requirements including inherited ones from parent communities */
  getEffectiveRequirements: async (communityId: string): Promise<EffectiveBadgeRequirement[]> => {
    const response = await apiClient.get(`/api/v1/communities/${communityId}/effective-requirements`)
    return response.data
  },
}

